src/resolver/descriptor.cc

```diff
@@ -0,0 +1,382 @@
+
+/**
+ * @brief Sets various fields on a `flox::resolver::ManifestDescriptor`
+ *        based on the `absPath` field.
+ * @param desc The descriptor to initialize.
+ * @param raw The raw description to parse.
+ */
+  static void
+initManifestDescriptorAbsPath(       ManifestDescriptor    & desc
+                             , const ManifestDescriptorRaw & raw
+                             )
+{
+  if ( ! raw.absPath.has_value() )
+    {
+      throw FloxException(
+        "`absPath' must be set when calling "
+        "`flox::resolver::ManifestDescriptor::initManifestDescriptorAbsPath'"
+       );
+    }
+
+  /* You might need to parse a globbed attr path, so handle that first. */
+  AttrPathGlob glob = maybeSplitAttrPathGlob( * raw.absPath );
+
+  if ( glob.size() < 3 )
+    {
+      throw FloxException(
+        "`absPath' must have at least three parts"
+      );
+    }
+
+  const auto & first = glob.front();
+  if ( ! first.has_value() )
+    {
+      throw FloxException(
+        "`absPath' may only have a glob as its second element"
+      );
+    }
+  desc.subtree = Subtree( * first );
+
+  if ( raw.stability.has_value() && ( first.value() != "catalog" ) )
+    {
+      throw FloxException(
+        "`stability' cannot be used with non-catalog paths"
+      );
+    }
+
+  if ( first.value() == "catalog" )
+    {
+      if ( glob.size() < 4 )
+        {
+          throw FloxException(
+            "`absPath' must have at least four parts for catalog paths"
+          );
+        }
+      const auto & third = glob.at( 2 );
+      if ( ! third.has_value() )
+        {
+          throw FloxException(
+            "`absPath' may only have a glob as its second element"
+          );
+        }
+      desc.stability = * third;
+      desc.path      = AttrPath {};
+      for ( auto itr = glob.begin() + 3; itr != glob.end(); ++itr )
+        {
+          const auto & elem = * itr;
+          if ( ! elem.has_value() )
+            {
+              throw FloxException(
+                "`absPath' may only have a glob as its second element"
+              );
+            }
+          desc.path->emplace_back( * elem );
+        }
+    }
+  else
+    {
+      desc.path = AttrPath {};
+      for ( auto itr = glob.begin() + 2; itr != glob.end(); ++itr )
+        {
+          const auto & elem = * itr;
+          if ( ! elem.has_value() )
+            {
+              throw FloxException(
+                "`absPath' may only have a glob as its second element"
+              );
+            }
+          desc.path->emplace_back( * elem );
+        }
+    }
+
+  const auto & second = glob.at( 1 );
+  if ( second.has_value() &&
+       ( ( * second ) != "null" ) && ( ( * second ) != "*" )
+     )
+    {
+      desc.systems = std::vector<std::string> { * second };
+      if ( raw.systems.has_value() && ( * raw.systems != * desc.systems ) )
+        {
+          throw FloxException(
+            "`systems' list conflicts with `absPath' system specification"
```

Comment:
There's a few places you're checking if values conflict. It seems like code/logic might be simpler to just forbid simultaneously setting any fields that could conflict. And I almost think that would be less confusing to use. Why would a user want to repeat themself?